### PR TITLE
Site Editor: patterns and templates cannot be edited from sidebar mobile view

### LINKED ISSUE #62649: Site Editor: Unable to select specific template on mobile
### Description

When navigating to the Site Editor using a mobile-sized viewport, it is not possible to tap templates (please see the screen recording below).

The issue can be reproduced with versions higher than `16.4.0`.

### Step-by-step reproduction instructions

1. Use a mobile-sized viewport.
2. In wp-admin, open the navigation, and open the site editor (Appearance → Editor).
3. Select "Templates"
4. Attempt to select a template.

### Screenshots, screen recording, code snippet

Here is the screen recording of the issue:

https://github.com/WordPress/gutenberg/assets/25105483/bf62463c-70d6-458d-a5e1-107b5ddfc806

### Environment info

Higher than `16.4.0`, e.g. `16.5.0` or `18.6.0-rc.1`

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### LINKED ISSUE #62932: Site Editor: "Edit Template" links unresponsive on mobile browsers (iPhone & Android)
### Description

Using the plugin (18.6.1) ellipsis buttons are unresponsive .

This happens with various mobile browsers but only on iPhone. Android works fine.

### Step-by-step reproduction instructions

1. Go to Appearance > Editor on an iPhone
2. Try clicking on any of the ellipsis buttons ⋮ and notice how they do not work.
3. On the Templates section, the drop-down menu does show, but the Edit button does nothing.

### Environment info

Gutenberg 18.6.1

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by ramonjd**
## What?

Fixes https://github.com/WordPress/gutenberg/issues/62932
Fixes https://github.com/WordPress/gutenberg/issues/62649

Specifically, this PR addressing the following:

1. Clicking on Edit from the Site Editor sidebar in "view" mode does not redirect to the editor.
2. Clicking on the WP logo once in edit mode crashes the browser

cc @oandregal to make sure everything from https://github.com/WordPress/gutenberg/pull/60689 still works

## Why?

So that clicking Edit on Templates and Patterns in the data views redirects the browser to the block editor in edit mode.

Also, to avoid crashing the browser.

## How?
In the mobile area prop in the Site Editor router, check for `canvas === 'edit'` and render the Editor component.

Skip using `document.startViewTransition` in mobile view when switching canvas modes.

## Testing Instructions

1. Fire up this branch and head to the Site Editor
3. Ensure your browser is narrower than `782px` wide to trigger the mobile view
4. Open the Templates sidebar panel
5. Click on a template or click on "Edit" in the ellipsis menu
6. The page should redirect to the editor
7. Click on the WP logo to return to "view" mode. The browser shouldn't crash. This primarily affected Chrome browsers for me, but it be related to the fact that `document.startViewTransition` doesn't have [full support outside Chrome](https://developer.mozilla.org/en-US/docs/Web/API/Document/startViewTransition).
8. Repeat the above for custom patterns and template parts

### Before

https://github.com/WordPress/gutenberg/assets/6458278/4d612d5c-ca98-467c-87ef-b923afd77c03

### After

https://github.com/WordPress/gutenberg/assets/6458278/a86453be-2a20-4795-abef-510904ef592f

### CODE REVIEW

**Review by andrewserong — COMMENTED** [review]
Thanks for looking into this! In general the links going deeper into page templates is working for me 👍

One issue I ran into, not sure if it's related, is that the W icon at the top left when attempting to go back from the edit view winds up crashing for me, with no error in the console:

https://github.com/WordPress/gutenberg/assets/14988353/92c67e99-a39b-4c36-8684-c3ceeb44582d

This doesn't happen for me when navigating up the hierarchy from editing a template. Only while editing a pattern. Can anyone reproduce that? Just wondering if it's something weird with my environment.

Not _quite_ related to this bug, but one thing I found confusing at first with the mobile version of the dataviews screens is that there's no W icon, so I had to rely on browser back buttons to navigate up the hierarchy. This is challenging to do when we add in the ability to go to the edit view on mobile, as the back button might take us "back" to the editor if we've come from there, so it can be difficult to navigate up to the root of the site editor on mobile if we're going into the edit view.

Here's a quick demo:

https://github.com/WordPress/gutenberg/assets/14988353/1850d809-d354-42a3-9e46-1bacc14fdfae

1. Go to Templates screen
2. Tap on a template to go to the edit view
3. Tap on the W icon to go up a level
4. There's no way to get out via tapping here, so use the browser back button
5. We wind up back in the editor again

**Comment by andrewserong** [comment]
> Repeat the above for custom patterns and template parts

Just checking if it's only me, but I wasn't able to find template parts from the mobile view. When I go to Patterns, I can only see filters for the synced or not synced patterns, and there doesn't appear to be links for the template parts like there is on Desktop 🤔 

image

**Comment by ramonjd** [comment]
Thanks for testing @andrewserong 🙇🏻 

So it sounds like the entire flow is broken for mobile. A bit of a shame, I was hoping for a quick win. I should know better 🤣 

I'll dig a little more deeply this week to suss out where things fit together and see what's fixable.

🍺

**Comment by t-hamano** [comment]
> One issue I ran into, not sure if it's related, is that the W icon at the top left when attempting to go back from the edit view winds up crashing for me, with no error in the console:

I was able to reproduce this issue. The easiest way is to go to the Pages screen, open any editor canvas, and click the W logo.

- This issue can be reproduced on any small viewport, regardless of whether it is a mobile device or not.
- Does not reproduce in WP6.6 RC1. Reproduces in Gutenberg trunk.

I think there are still many issues with the mobile viewport. The main issue is probably #62878.

It's a difficult question as to what extent this needs to be fixed in WP6.6 😅

**Comment by ramonjd** [comment]
> It's a difficult question as to what extent this needs to be fixed in WP6.6 

Thanks for the thoughts @t-hamano 

After thinking about it for a nanosecond, I'll put this out there. For 6.6, users should at least be able to:

- edit templates et al. in narrow viewport width 
- navigate back to the "view" sidebar

So I'll investigate the crash. With those things fixed hopefully we've restored base functionality. 

Any UX cluster bombs we can fix as follow ups.

**Comment by ramonjd** [comment]
> Just checking if it's only me, but I wasn't able to find template parts from the mobile view. When I go to Patterns, I can only see filters for the synced or not synced patterns, and there doesn't appear to be links for the template parts like there is on Desktop

Forgot to reply to this, sorry.

I don't see any filters for template parts on Desktop either. 

Screenshot 2024-07-02 at 11 12 43 AM

It looks like Template Parts are accessible also from the sidebar view, but in mobile view, this sidebar is not accessible so there's no way for users to list them. 😅

**Comment by ramonjd** [comment]
> So I'll investigate the crash. With those things fixed hopefully we've restored base functionality.

I think I've traced the crash to the use of `document.startViewTransition` in [setCanvasMode](https://github.com/WordPress/gutenberg/blob/1a8c9485ca364ed9c7b9a35b544332613be18254/packages/edit-site/src/store/private-actions.js#L13-L13). 

Only firing that in desktop mode seems to have fixed the crashing for me, however I haven't located the root cause. Chrome supports `document.startViewTransition` - I expect it short circuits rendering somehow as the entire HTML document is kaputt. Not sure.

**Comment by ramonjd** [comment]
> I found confusing at first with the mobile version of the dataviews screens is that there's no W icon, so I had to rely on browser back buttons to navigate up the hierarchy. 

There's an issue for this: https://github.com/WordPress/gutenberg/issues/62878 

I hacked together something to address it in the short term, and have commented over there. 👍🏻

File: packages/edit-site/src/store/private-actions.js [hunk 1]
```diff
@@ -13,15 +13,15 @@ import { store as editorStore } from '@wordpress/editor';
...
 		const switchCanvasMode = () => {
 			registry.batch( () => {
-				const isMediumOrBigger =
-					window.matchMedia( '(min-width: 782px)' ).matches;
 				registry.dispatch( blockEditorStore ).clearSelectedBlock();
 				registry.dispatch( editorStore ).setDeviceType( 'Desktop' );
 				registry
 					.dispatch( blockEditorStore )
-					.__unstableSetEditorMode( 'edit' );
+					.__unstableSetEditorMode( mode );
```
Comment (youknowriad) [inline packages/edit-site/src/store/private-actions.js]: I'm not sure this change is correct. There's no "view" mode in the block editor mode.

Comment (ramonjd) [reply packages/edit-site/src/store/private-actions.js]: Thanks for catching that. That was a hangover from debugging 🤦🏻 

I'll update quickly

**Review by youknowriad — APPROVED** [review]
LGTM 👍

**Comment by ramonjd** [comment]
Thanks for the reviews, folks!

**Comment by ramonjd** [comment]
Cherry-pick and merge to the wp/6.6 branch here:

- https://github.com/WordPress/gutenberg/pull/63055

**Comment by oandregal** [comment]
I've prepared https://github.com/WordPress/gutenberg/pull/63060 as a potential quick change for 6.6.

**Pull request merged**